File: apps/ios/Sources/Views/AddServerSheet.swift
import SwiftUI

/// Add-Server entry point. Replaces the three side-by-side entry
/// buttons in the old Settings form with a focused four-card sheet:
/// Scan QR · Discover on LAN · SSH bootstrap · Paste URL+token.
///
/// Each card pushes (or sheets) into a specialised flow; once that
/// flow finishes setting the endpoint, this sheet auto-dismisses.
struct AddServerSheet: View {
    @Environment(SessionStore.self) private var store
    @Environment(\.dismiss) private var dismiss
    @Environment(\.colorScheme) private var colorScheme

    @State private var showScanner = false
    @State private var showDiscover = false
    @State private var showSshLogin = false
    @State private var showManual = false

    /// We watch this so manual entry can auto-dismiss after success.
    private var lastEndpoint: String { store.endpoint.url }

    var body: some View {
        NavigationStack {
            ZStack {
                SweKittyTheme.backgroundGradient(for: colorScheme).ignoresSafeArea()
                ScrollView {
                    VStack(spacing: 14) {
                        intro
                        entryCard(
                            icon: "qrcode.viewfinder",
                            tint: SweKittyTheme.accentStrong,
                            title: "Scan pairing QR",
                            subtitle: "Camera-scan the QR from the broker terminal.",
                            action: { showScanner = true }
                        )
                        entryCard(
                            icon: "wifi.circle",
                            tint: SweKittyTheme.codexAccent,
                            title: "Discover on LAN",
                            subtitle: "Find a broker advertising via mDNS on the same Wi-Fi.",
                            action: { showDiscover = true }
                        )
                        entryCard(
                            icon: "terminal",
                            tint: SweKittyTheme.claudeAccent,
                            title: "SSH bootstrap",
                            subtitle: "Cold-start a broker on a remote box you can SSH to.",
                            action: { showSshLogin = true }
                        )
                        entryCard(
                            icon: "link",
                            tint: SweKittyTheme.warning,
                            title: "Paste URL + token",
                            subtitle: "If you already have ws://… + the bearer token.",
                            action: { showManual = true }
                        )
                    }
                    .padding(.horizontal, 16)
                    .padding(.vertical, 18)
                }
                .scrollIndicators(.hidden)
            }
            .navigationTitle("Add server")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Cancel") { dismiss() }
                }
            }
            .sheet(isPresented: $showScanner) {
                QRScannerSheet { code in
                    if let parsed = PairingURL.parse(code) {
                        let next = StoredEndpoint(url: parsed.endpoint, token: parsed.token)
                        store.endpoint = next
                        store.upsertSavedServer(name: next.displayHost, endpoint: next, makeDefault: true)
                        store.disconnect()
                        store.connect()
                        dismiss()
                    }
                }
            }
            .sheet(isPresented: $showDiscover) {
                DiscoveryView()
                    .environment(store)
                    .onDisappear { if store.endpoint.isComplete { dismiss() } }
            }
            .sheet(isPresented: $showSshLogin) {
                SSHLoginSheet()
                    .environment(store)
                    .onDisappear { if store.endpoint.isComplete { dismiss() } }
            }
            .sheet(isPresented: $showManual) {
                ManualPairSheet()
                    .environment(store)
                    .onDisappear { if store.endpoint.isComplete { dismiss() } }
            }
        }
        .tint(SweKittyTheme.accentStrong)
    }

    private var intro: some View {
        VStack(alignment: .leading, spacing: 6) {
            Text("Where is the broker?")
                .font(.headline)
                .foregroundStyle(SweKittyTheme.textPrimary)
            Text("Pick how this device should reach the swe-kitty server. You can change servers later.")
                .font(.subheadline)
                .foregroundStyle(SweKittyTheme.textSecondary)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal, 14)
        .padding(.vertical, 12)
        .glassRoundedRect()
    }

    private func entryCard(icon: String, tint: Color, title: String, subtitle: String, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            HStack(spacing: 14) {
                Image(systemName: icon)
                    .font(.title2)
                    .foregroundStyle(SweKittyTheme.textOnAccent)
                    .frame(width: 42, height: 42)
                    .background(tint)
                    .clipShape(Circle())
                VStack(alignment: .leading, spacing: 2) {
                    Text(title)
                        .font(.subheadline.weight(.semibold))
                        .foregroundStyle(SweKittyTheme.textPrimary)
                    Text(subtitle)
                        .font(.caption)
                        .foregroundStyle(SweKittyTheme.textMuted)
                        .lineLimit(2)
                }
                Spacer(minLength: 8)
                Image(systemName: "chevron.right")
                    .font(.caption.weight(.semibold))
                    .foregroundStyle(SweKittyTheme.textMuted)
            }
            .padding(.horizontal, 14)
            .padding(.vertical, 12)
            .glassRect(cornerRadius: SweKittyTheme.cardCornerRadius, tint: tint.opacity(0.16))
        }
        .buttonStyle(.plain)
    }
}

/// Plain URL+token manual-entry sheet, extracted from the old
/// `SettingsSheet.pairingCard`. Only the bare-minimum fields — no
/// Browse-directory, no Connect+Start triple. After saving, the
/// post-pair agent picker handles the "start a session" step.
struct ManualPairSheet: View {
    @Environment(SessionStore.self) private var store
    @Environment(\.dismiss) private var dismiss
    @Environment(\.colorScheme) private var colorScheme

    @State private var url: String = ""
    @State private var token: String = ""
    @State private var error: String?

    var body: some View {
        NavigationStack {
            ZStack {
                SweKittyTheme.backgroundGradient(for: colorScheme).ignoresSafeArea()
                ScrollView {
                    VStack(spacing: 14) {
                        VStack(alignment: .leading, spacing: 10) {
                            Text("Harness URL")
                                .font(.caption.weight(.semibold))
                                .foregroundStyle(SweKittyTheme.textSecondary)
                            TextField("ws://192.168.1.10:1977", text: $url)
                                .textInputAutocapitalization(.never)
                                .autocorrectionDisabled()
                                .keyboardType(.URL)
                                .textFieldStyle(.plain)
                                .padding(10)
                                .background(SweKittyTheme.surface.opacity(0.6))
                                .clipShape(RoundedRectangle(cornerRadius: 12, style: .continuous))

                            Text("Bearer token")
                                .font(.caption.weight(.semibold))
                                .foregroundStyle(SweKittyTheme.textSecondary)
                            SecureField("Bearer token", text: $token)
                                .textInputAutocapitalization(.never)
                                .autocorrectionDisabled()
                                .textFieldStyle(.plain)
                                .padding(10)
                                .background(SweKittyTheme.surface.opacity(0.6))
                                .clipShape(RoundedRectangle(cornerRadius: 12, style: .continuous))

                            if let error {
                                Text(error)
                                    .font(.footnote)
                                    .foregroundStyle(SweKittyTheme.danger)
                            }
                        }
                        .padding(14)
                        .glassRoundedRect()

                        Button(action: save) {
                            Label("Save & connect", systemImage: "link")
                                .font(.subheadline.weight(.semibold))
                                .foregroundStyle(SweKittyTheme.textPrimary)
                                .frame(maxWidth: .infinity)
                                .padding(.vertical, 10)
                                .glassCapsule(interactive: true, tint: SweKittyTheme.accentStrong.opacity(0.55))
                        }
                        .buttonStyle(.plain)
                        .disabled(url.isEmpty || token.isEmpty)
                    }
                    .padding(.horizontal, 16)
                    .padding(.vertical, 18)
                }
                .scrollIndicators(.hidden)
            }
            .navigationTitle("Paste URL + token")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Cancel") { dismiss() }
                }
            }
            .onAppear {
                url = store.endpoint.url
                token = store.endpoint.token
            }
        }
        .tint(SweKittyTheme.accentStrong)
    }

    private func save() {
        let trimmedUrl = url.trimmingCharacters(in: .whitespaces)
        let trimmedToken = token.trimmingCharacters(in: .whitespaces)
        guard !trimmedUrl.isEmpty, !trimmedToken.isEmpty else {
            error = "URL and token are both required."
            return
        }
        let next = StoredEndpoint(url: trimmedUrl, token: trimmedToken)
        store.endpoint = next
        store.upsertSavedServer(name: next.displayHost, endpoint: next, makeDefault: true)
        store.disconnect()
        store.connect()
        dismiss()
    }
}

File: apps/android/app/src/main/kotlin/sh/nikhil/swekitty/ui/AddServerSheet.kt
package sh.nikhil.swekitty.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ChevronRight
import androidx.compose.material.icons.filled.Link
import androidx.compose.material.icons.filled.QrCodeScanner
import androidx.compose.material.icons.filled.Terminal
import androidx.compose.material.icons.filled.Wifi
import androidx.compose.material3.Button
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import sh.nikhil.swekitty.Endpoint
import sh.nikhil.swekitty.PairingURL
import sh.nikhil.swekitty.SessionStore

/**
 * Compose mirror of `apps/ios/Sources/Views/AddServerSheet.swift`.
 * Four entry-method cards — Scan QR / Discover LAN / SSH / Manual.
 * Replaces the giant pairing form in Settings.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun AddServerSheet(store: SessionStore, onDismiss: () -> Unit) {
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    var showScanner by remember { mutableStateOf(false) }
    var showDiscover by remember { mutableStateOf(false) }
    var showSsh by remember { mutableStateOf(false) }
    var showManual by remember { mutableStateOf(false) }

    val endpoint by store.endpoint.collectAsState()

    /**
     * Common landing for the QR scanner sheet. Returns true when the
     * code was a valid pairing URL and the sheet should close.
     */
    fun applyPairingCode(code: String?): Boolean {
        if (code.isNullOrBlank()) return false
        val parsed = PairingURL.parse(code) ?: return false
        val ep = Endpoint(parsed.endpoint, parsed.token)
        store.setEndpoint(ep.url, ep.token)
        store.upsertSavedServer(ep.displayHost, ep, makeDefault = true)
        store.disconnect()
        store.connect()
        return true
    }

    ModalBottomSheet(onDismissRequest = onDismiss, sheetState = sheetState) {
        Column(
            modifier = Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 12.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            Text("Add server", style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.SemiBold)
            Text(
                "Pick how this device should reach the swe-kitty server. You can switch servers later from Settings.",
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            EntryCard(
                icon = { Icon(Icons.Filled.QrCodeScanner, null, tint = Color.White) },
                tint = SweKittyTheme.accentStrong(),
                title = "Scan pairing QR",
                subtitle = "Camera or pick from Photos.",
            ) { showScanner = true }
            EntryCard(
                icon = { Icon(Icons.Filled.Wifi, null, tint = Color.White) },
                tint = SweKittyTheme.codexAccent(),
                title = "Discover on LAN",
                subtitle = "Find a broker advertising via mDNS on the same Wi-Fi.",
            ) { showDiscover = true }
            EntryCard(
                icon = { Icon(Icons.Filled.Terminal, null, tint = Color.White) },
                tint = SweKittyTheme.claudeAccent(),
                title = "SSH bootstrap",
                subtitle = "Cold-start a broker on a remote box you can SSH to.",
            ) { showSsh = true }
            EntryCard(
                icon = { Icon(Icons.Filled.Link, null, tint = Color.White) },
                tint = SweKittyTheme.warning(),
                title = "Paste URL + token",
                subtitle = "If you already have ws://… + a bearer token.",
            ) { showManual = true }
            Spacer(Modifier.height(8.dp))
        }
    }

    if (showScanner) {
        // Full-screen dialog so the camera preview owns the whole
        // window — matches the iOS QRScannerSheet presentation. The
        // sheet itself handles camera permission and the gallery
        // fallback, so the host only feeds the resulting code through
        // the shared pairing handler.
        Dialog(
            onDismissRequest = { showScanner = false },
            properties = DialogProperties(usePlatformDefaultWidth = false),
        ) {
            QRScannerSheet(
                onScanned = { code ->
                    showScanner = false
                    if (applyPairingCode(code)) onDismiss()
                },
                onDismiss = { showScanner = false },
            )
        }
    }
    if (showDiscover) {
        DiscoveryScreen(store, onDismiss = {
            showDiscover = false
            if (endpoint.isComplete) onDismiss()
        })
    }
    if (showSsh) {
        SSHLoginSheet(store, onDismiss = {
            showSsh = false
            if (endpoint.isComplete) onDismiss()
        })
    }
    if (showManual) {
        ManualPairSheet(store, onDismiss = {
            showManual = false
            if (endpoint.isComplete) onDismiss()
        })
    }
}

@Composable
private fun EntryCard(
    icon: @Composable () -> Unit,
    tint: Color,
    title: String,
    subtitle: String,
    onTap: () -> Unit,
) {
    Surface(
        shape = RoundedCornerShape(SweKittyTheme.cardCornerRadiusDp.dp),
        color = tint.copy(alpha = 0.12f),
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(SweKittyTheme.cardCornerRadiusDp.dp))
            .clickable(onClick = onTap),
    ) {
        Row(
            modifier = Modifier.fillMaxWidth().padding(14.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Box(
                modifier = Modifier.size(42.dp).clip(CircleShape).background(tint),
                contentAlignment = Alignment.Center,
            ) { icon() }
            Spacer(Modifier.width(14.dp))
            Column(modifier = Modifier.weight(1f)) {
                Text(title, style = MaterialTheme.typography.titleSmall, fontWeight = FontWeight.SemiBold)
                Text(subtitle, style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
            }
            Icon(Icons.Filled.ChevronRight, null, tint = MaterialTheme.colorScheme.onSurfaceVariant)
        }
    }
}

/**
 * Plain URL+token manual entry — extracted from the old SettingsScreen.
 * Only the bare-minimum fields; the post-pair agent picker covers
 * "start a session" so no Connect+Start triple here.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ManualPairSheet(store: SessionStore, onDismiss: () -> Unit) {
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    val endpoint by store.endpoint.collectAsState()
    var url by remember { mutableStateOf(endpoint.url) }
    var token by remember { mutableStateOf(endpoint.token) }

    ModalBottomSheet(onDismissRequest = onDismiss, sheetState = sheetState) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 16.dp, vertical = 12.dp)
                .verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            Text("Paste URL + token", style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.SemiBold)
            OutlinedTextField(
                value = url,
                onValueChange = { url = it },
                label = { Text("ws:// URL") },
                singleLine = true,
                modifier = Modifier.fillMaxWidth(),
            )
            OutlinedTextField(
                value = token,
                onValueChange = { token = it },
                label = { Text("Bearer token") },
                singleLine = true,
                modifier = Modifier.fillMaxWidth(),
            )
            Button(
                onClick = {
                    val ep = Endpoint(url.trim(), token.trim())
                    if (ep.isComplete) {
                        store.setEndpoint(ep.url, ep.token)
                        store.upsertSavedServer(ep.displayHost, ep, makeDefault = true)
                        store.disconnect()
                        store.connect()
                        onDismiss()
                    }
                },
                enabled = url.isNotBlank() && token.isNotBlank(),
                modifier = Modifier.fillMaxWidth(),
            ) { Text("Save & connect") }
            Spacer(Modifier.height(8.dp))
        }
    }
}
